File: src/test/java/com/joecollins/graphics/components/lowerthird/LowerThirdHeadlineAndSummarySingleLabelTest.java
package com.joecollins.graphics.components.lowerthird;

import static com.joecollins.graphics.utils.RenderTestUtils.compareRendering;
import static org.junit.Assert.assertEquals;

import com.joecollins.bindings.Binding;
import com.joecollins.bindings.IndexedBinding;
import com.joecollins.graphics.components.lowerthird.SummaryWithoutLabels.Entry;
import java.awt.Color;
import java.io.IOException;
import java.time.Clock;
import java.time.Instant;
import java.time.ZoneId;
import org.junit.Test;

public class LowerThirdHeadlineAndSummarySingleLabelTest {

  @Test
  public void testHeadline() {
    LowerThirdHeadlineAndSummarySingleLabel lowerThird =
        new LowerThirdHeadlineAndSummarySingleLabel();
    lowerThird.setHeadlineBinding(Binding.fixedBinding("POLLS CLOSE ACROSS CENTRAL CANADA"));
    assertEquals("POLLS CLOSE ACROSS CENTRAL CANADA", lowerThird.getHeadline());
  }

  @Test
  public void testSubhead() {
    LowerThirdHeadlineAndSummarySingleLabel lowerThird =
        new LowerThirdHeadlineAndSummarySingleLabel();
    lowerThird.setSubheadBinding(Binding.fixedBinding("Polls open for 30 minutes on west coast"));
    assertEquals("Polls open for 30 minutes on west coast", lowerThird.getSubhead());
  }

  @Test
  public void testSummaryPanel() {
    LowerThirdHeadlineAndSummarySingleLabel lowerThird =
        new LowerThirdHeadlineAndSummarySingleLabel();
    lowerThird.setNumSummaryEntriesBinding(Binding.fixedBinding(2));
    lowerThird.setSummaryHeaderBinding(() -> "170 SEATS FOR MAJORITY");
    lowerThird.setSummaryEntriesBinding(
        IndexedBinding.listBinding(new Entry(Color.RED, "2"), new Entry(Color.BLUE, "1")));
    assertEquals(2, lowerThird.getNumSummaryEntries());

    assertEquals(Color.RED, lowerThird.getEntryColor(0));
    assertEquals(Color.BLUE, lowerThird.getEntryColor(1));

    assertEquals("2", lowerThird.getEntryValue(0));
    assertEquals("1", lowerThird.getEntryValue(1));

    assertEquals("170 SEATS FOR MAJORITY", lowerThird.getSummaryHeader());
  }

  @Test
  public void testRenderHeadlineAndSummary() throws IOException {
    LowerThirdHeadlineAndSummarySingleLabel lowerThird =
        new LowerThirdHeadlineAndSummarySingleLabel();
    lowerThird.setSize(1024, 50);
    lowerThird.setLeftImageBinding(
        Binding.fixedBinding(
            LowerThird.createImage(
                LowerThirdHeadlineAndSummarySingleLabelTest.class
                    .getClassLoader()
                    .getResource("com/joecollins/graphics/lowerthird-left.png"))));
    lowerThird.setPlaceBinding(Binding.fixedBinding("OTTAWA"));
    lowerThird.setTimeZoneBinding(Binding.fixedBinding(ZoneId.of("Canada/Eastern")));
    lowerThird.setClock(Clock.fixed(Instant.parse("2019-10-22T01:30:00Z"), ZoneId.systemDefault()));
    lowerThird.setHeadlineBinding(Binding.fixedBinding("CENTRAL CANADA POLLS CLOSE"));
    lowerThird.setSubheadBinding(Binding.fixedBinding("Polls open for 30 minutes on west coast"));
    lowerThird.setNumSummaryEntriesBinding(Binding.fixedBinding(2));
    lowerThird.setSummaryEntriesBinding(
        IndexedBinding.listBinding(new Entry(Color.RED, "2"), new Entry(Color.BLUE, "1")));
    lowerThird.setSummaryHeaderBinding(Binding.fixedBinding("170 SEATS FOR MAJORITY"));

    compareRendering("LowerThird", "HeadlineAndSummarySingleLabel", lowerThird);
  }

  @Test
  public void testRenderLongHeadlineAndSummary() throws IOException {
    LowerThirdHeadlineAndSummarySingleLabel lowerThird =
        new LowerThirdHeadlineAndSummarySingleLabel();
    lowerThird.setSize(1024, 50);
    lowerThird.setLeftImageBinding(
        Binding.fixedBinding(
            LowerThird.createImage(
                LowerThirdHeadlineAndSummarySingleLabelTest.class
                    .getClassLoader()
                    .getResource("com/joecollins/graphics/lowerthird-left.png"))));
    lowerThird.setPlaceBinding(Binding.fixedBinding("OTTAWA"));
    lowerThird.setTimeZoneBinding(Binding.fixedBinding(ZoneId.of("Canada/Eastern")));
    lowerThird.setClock(Clock.fixed(Instant.parse("2019-10-22T01:30:00Z"), ZoneId.systemDefault()));
    lowerThird.setHeadlineBinding(Binding.fixedBinding("POLLS CLOSE ACROSS CENTRAL CANADA"));
    lowerThird.setSubheadBinding(
        Binding.fixedBinding("Polls open for another 30 minutes in British Columbia, Yukon"));
    lowerThird.setNumSummaryEntriesBinding(Binding.fixedBinding(2));
    lowerThird.setSummaryEntriesBinding(
        IndexedBinding.listBinding(new Entry(Color.RED, "2"), new Entry(Color.BLUE, "1")));
    lowerThird.setSummaryHeaderBinding(Binding.fixedBinding("170 SEATS FOR MAJORITY"));

    compareRendering("LowerThird", "LongHeadlineAndSummarySingleLabel", lowerThird);
  }
}

File: src/test/kotlin/com/joecollins/graphics/components/lowerthird/LowerThirdHeadlineAndSummarySingleLabelTest.kt
package com.joecollins.graphics.components.lowerthird

import com.joecollins.graphics.components.lowerthird.LowerThird.Companion.createImage
import com.joecollins.graphics.utils.RenderTestUtils.compareRendering
import com.joecollins.pubsub.asOneTimePublisher
import org.awaitility.Awaitility
import org.hamcrest.core.IsEqual
import org.junit.Assert
import org.junit.Test
import java.awt.Color
import java.io.IOException
import java.time.Clock
import java.time.Instant
import java.time.ZoneId
import java.util.concurrent.TimeUnit

class LowerThirdHeadlineAndSummarySingleLabelTest {
    @Test
    fun testHeadline() {
        val lowerThird = LowerThirdHeadlineAndSummarySingleLabel(
            leftImagePublisher = createImage("", Color.RED, Color.WHITE).asOneTimePublisher(),
            placePublisher = "".asOneTimePublisher(),
            timezonePublisher = ZoneId.systemDefault().asOneTimePublisher(),
            headlinePublisher = "POLLS CLOSE ACROSS CENTRAL CANADA".asOneTimePublisher(),
            subheadPublisher = "Polls open for 30 minutes on west coast".asOneTimePublisher(),
            summaryHeaderPublisher = "170 SEATS FOR MAJORITY".asOneTimePublisher(),
            summaryEntriesPublisher =
            listOf(
                SummaryWithoutLabels.Entry(Color.RED, "2"),
                SummaryWithoutLabels.Entry(Color.BLUE, "1")
            )
                .asOneTimePublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ lowerThird.headline }, IsEqual("POLLS CLOSE ACROSS CENTRAL CANADA"))
    }

    @Test
    fun testSubhead() {
        val lowerThird = LowerThirdHeadlineAndSummarySingleLabel(
            leftImagePublisher = createImage("", Color.RED, Color.WHITE).asOneTimePublisher(),
            placePublisher = "".asOneTimePublisher(),
            timezonePublisher = ZoneId.systemDefault().asOneTimePublisher(),
            headlinePublisher = "POLLS CLOSE ACROSS CENTRAL CANADA".asOneTimePublisher(),
            subheadPublisher = "Polls open for 30 minutes on west coast".asOneTimePublisher(),
            summaryHeaderPublisher = "170 SEATS FOR MAJORITY".asOneTimePublisher(),
            summaryEntriesPublisher =
            listOf(
                SummaryWithoutLabels.Entry(Color.RED, "2"),
                SummaryWithoutLabels.Entry(Color.BLUE, "1")
            )
                .asOneTimePublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ lowerThird.subhead }, IsEqual("Polls open for 30 minutes on west coast"))
    }

    @Test
    fun testSummaryPanel() {
        val lowerThird = LowerThirdHeadlineAndSummarySingleLabel(
            leftImagePublisher = createImage("", Color.RED, Color.WHITE).asOneTimePublisher(),
            placePublisher = "".asOneTimePublisher(),
            timezonePublisher = ZoneId.systemDefault().asOneTimePublisher(),
            headlinePublisher = "POLLS CLOSE ACROSS CENTRAL CANADA".asOneTimePublisher(),
            subheadPublisher = "Polls open for 30 minutes on west coast".asOneTimePublisher(),
            summaryHeaderPublisher = "170 SEATS FOR MAJORITY".asOneTimePublisher(),
            summaryEntriesPublisher =
            listOf(
                SummaryWithoutLabels.Entry(Color.RED, "2"),
                SummaryWithoutLabels.Entry(Color.BLUE, "1")
            )
                .asOneTimePublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ lowerThird.numSummaryEntries }, IsEqual(2))
        Assert.assertEquals(Color.RED, lowerThird.getEntryColor(0))
        Assert.assertEquals(Color.BLUE, lowerThird.getEntryColor(1))
        Assert.assertEquals("2", lowerThird.getEntryValue(0))
        Assert.assertEquals("1", lowerThird.getEntryValue(1))
        Assert.assertEquals("170 SEATS FOR MAJORITY", lowerThird.summaryHeader)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderHeadlineAndSummary() {
        val lowerThird = LowerThirdHeadlineAndSummarySingleLabel(
            leftImagePublisher =
            createImage(
                LowerThirdHeadlineAndSummarySingleLabelTest::class.java
                    .classLoader
                    .getResource("com/joecollins/graphics/lowerthird-left.png")
            )
                .asOneTimePublisher(),
            placePublisher = "OTTAWA".asOneTimePublisher(),
            timezonePublisher = ZoneId.of("Canada/Eastern").asOneTimePublisher(),
            headlinePublisher = "CENTRAL CANADA POLLS CLOSE".asOneTimePublisher(),
            subheadPublisher = "Polls open for 30 minutes on west coast".asOneTimePublisher(),
            summaryHeaderPublisher = "170 SEATS FOR MAJORITY".asOneTimePublisher(),
            summaryEntriesPublisher =
            listOf(
                SummaryWithoutLabels.Entry(Color.RED, "2"),
                SummaryWithoutLabels.Entry(Color.BLUE, "1")
            )
                .asOneTimePublisher(),
            clock = Clock.fixed(Instant.parse("2019-10-22T01:30:00Z"), ZoneId.systemDefault())
        )
        lowerThird.setSize(1024, 50)
        compareRendering("LowerThird", "HeadlineAndSummarySingleLabel", lowerThird)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderLongHeadlineAndSummary() {
        val lowerThird = LowerThirdHeadlineAndSummarySingleLabel(
            leftImagePublisher =
            createImage(
                LowerThirdHeadlineAndSummarySingleLabelTest::class.java
                    .classLoader
                    .getResource("com/joecollins/graphics/lowerthird-left.png")
            )
                .asOneTimePublisher(),
            placePublisher = "OTTAWA".asOneTimePublisher(),
            timezonePublisher = ZoneId.of("Canada/Eastern").asOneTimePublisher(),
            headlinePublisher = "POLLS CLOSE ACROSS CENTRAL CANADA".asOneTimePublisher(),
            subheadPublisher = "Polls open for another 30 minutes in British Columbia, Yukon".asOneTimePublisher(),
            summaryHeaderPublisher = "170 SEATS FOR MAJORITY".asOneTimePublisher(),
            summaryEntriesPublisher =
            listOf(
                SummaryWithoutLabels.Entry(Color.RED, "2"),
                SummaryWithoutLabels.Entry(Color.BLUE, "1")
            )
                .asOneTimePublisher(),
            clock = Clock.fixed(Instant.parse("2019-10-22T01:30:00Z"), ZoneId.systemDefault())
        )
        lowerThird.setSize(1024, 50)
        compareRendering("LowerThird", "LongHeadlineAndSummarySingleLabel", lowerThird)
    }
}